cps/src/main/java/org/jenkinsci/plugins/workflow/cps/rerun/RerunAction.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.cps.rerun;
+
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.CauseAction;
+import hudson.model.Item;
+import hudson.model.Job;
+import hudson.model.Run;
+import hudson.model.queue.QueueTaskFuture;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import javax.servlet.ServletException;
+import jenkins.model.Jenkins;
+import jenkins.model.ParameterizedJobMixIn;
+import jenkins.model.TransientActionFactory;
+import jenkins.scm.api.SCMRevisionAction;
+import org.acegisecurity.AccessDeniedException;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+/**
+ * Attached to a {@link WorkflowRun} when it could be rerun.
+ */
+public class RerunAction implements Action {
+
+    private final Run run;
+
+    private RerunAction(Run run) {
+        this.run = run;
+    }
+
+    @Override public String getDisplayName() {
+        return "Rerun with Edits";
+    }
+
+    @Override public String getIconFileName() {
+        return isEnabled() ? "clock.png" : null;
+    }
+
+    @Override public String getUrlName() {
+        return isEnabled() ? "rerun" : null;
+    }
+
+    private @CheckForNull CpsFlowExecution getExecution() {
+        FlowExecution exec = ((FlowExecutionOwner.Executable) run).asFlowExecutionOwner().getOrNull();
+        return exec instanceof CpsFlowExecution ? (CpsFlowExecution) exec : null;
+    }
+
+    /* accessible to Jelly */ public boolean isEnabled() {
+        Job job = run.getParent();
+        if (!job.hasPermission(Item.BUILD)) {
+            return false;
+        }
+        if (!job.hasPermission(Item.EXTENDED_READ)) {
+            // CONFIGURE would fail on a branch project, and anyway we are not changing the job configuration per se.
+            // We do want to be able to see the script (if CpsFlowDefinition).
+            // TODO but for multibranch we arguably want to limit this to people who could have changed the MultiBranchProject or OrganizationFolder.
+            // Do we need a new permission just for this action?
```

Comment:
As written, this is probably a security hole. Say someone created a job using folder-scoped credentials to do some sensitive deployment, then granted `EXTENDED_READ` on the folder to others (which lets them see the name of the credentials, but of course not the password, plus the job configuration generally). Allowing them to rerun with an edited script would allow them to hijack the credentials.

Ideally for `CpsScmFlowDefinition` and multibranch the requisite permission would match a user’s ability to actually commit to the selected SCM: in other words, to make such an edit persistent. But Jenkins has no way of checking that directly (short of using the `github-oauth` plugin or similar). We could check whether the `User` has recorded commits in the job’s changelog, though this seems easily spoofed (for example by editing your user profile to include someone else’s email address). The only foolproof and generic way to determine if someone _could_ commit to a given SCM repository (and branch, which may have its own ACL) is to challenge them to do so: display a random token and ask them to commit it to some file. This seems overly burdensome, and it is unclear how you would cache this approval in a way that allows it to be retracted if SCM permissions are changed.

Enforcing `CONFIGURE` at the `WorkflowJob` level would prevent this from being used under any circumstances on multibranch projects, so that is a nonstarter. Might suffice to check for `CONFIGURE` at this level _or above_, which would then give a pass to anyone who would anyway be able to edit the `WorkflowMultiBranchProject` or `OrganizationFolder`. On the other hand it would block non-admins from using the feature in a system which is 100% pipeline-as-code (only `READ` + `BUILD` for developers).

A new permission would work, but adds to the `AuthorizationStrategy` configuration UI, and imposes a burden on installations using this feature to grant it. Perhaps it could be `impliedBy(CONFIGURE)`, which I *think* would result in it being granted even on a branch project if the user had `CONFIGURE` at the multibranch or higher level in the `AuthorizationStrategy`. Tricky because it depends on the order of interpretation of the `WorkflowJob.getACL` override vs. `Permission.impliedBy` search in the `AuthorizationStrategy`: the blocking of `CONFIGURE` & `DELETE` on branch projects and organization folder children is done “at the last minute” (invisible to the `AuthorizationStrategy`).

@stephenc might have some advice.